### PR TITLE
Regression test for doctest reporting

### PR DESCRIPTION
**Pull request by Zac-HD**
This pull replaces #4243, and closes issue #4223. 

Well, mostly - it turns out the underlying issue has been fixed on the master branch in the last four months, so really all I'm adding is the regression tests I wrote and related code cleanup.  CC  @tk0miya - this should be much faster to review than expected!

I have added the reporting issue to the changelog, even though this pull is no longer what fixed it, and my name to the authors list for the tests (hopefully that's enough).

### CODE REVIEW

**Comment by Zac-HD** [comment]
Oh dear - the underlying problem is fixed in my local environment, but *not* in CI.  I think I'll need to reapply the patches, which means checking again to see if I can fix the underlying issue instead of applying patches drawn from #4243.  ETA another day or two - let me know if the schedule is tight and I'll prioritise.

**Comment by tk0miya** [comment]
It seems tests are failed. Could you check this please?

```
E           assert 'File "dir/inner.rst", line 1, in default' in ['File "dir/inner.rst", line 2, in default', 'File "dir/inner.rst", line 2, in default', 'File "index.rst", line 2, in default', 'File "index.rst", line 5, in default']
```
https://travis-ci.org/sphinx-doc/sphinx/jobs/328654000#L1801

Note: I think this is not a new feature, so we can merge this after feature freeze.

**Comment by Zac-HD** [comment]
Yeah, that's the bug - the doctests are actually in four files not two, and not at those line numbers in any case 😭

Next step, I'll try again to reproduce locally...  see my comment above.

**Comment by tk0miya** [comment]
Oh, I'd overlooked your last comment. I can't say X-day at this moment because I don't see status of other issues now. But I think this is not a new feature and it seems not breaking changes.
So please go at your pace.

File: sphinx/environment/__init__.py [hunk 1]
```diff
@@ -358,17 +357,16 @@ def merge_info_from(self, docnames, other, app):
...
         """
         if filename.startswith(self.srcdir):
-            filename = filename[len(self.srcdir) + 1:]
+            filename = os.path.relpath(filename, self.srcdir)
         for suffix in self.config.source_suffix:
-            if fnmatch.fnmatch(filename, '*' + suffix):
-                return filename[:-len(suffix)]
-        # the file does not have docname
+            if filename.endswith(suffix):
+                return filename.rstrip(suffix)
```
Comment (tk0miya) [inline sphinx/environment/__init__.py]: I noticed this is not good. `str.rstrip()` is not alternative of `filename[:-len(suffix]`.
```
>>> "rrrr.rst".rstrip(".rst")
''
```

File: sphinx/environment/__init__.py [hunk 2]
```diff
@@ -358,17 +357,16 @@ def merge_info_from(self, docnames, other, app):
         app.emit('env-merge-info', self, docnames, other)
 
     def path2doc(self, filename):
-        # type: (unicode) -> unicode
+        # type: (unicode) -> Optional[unicode]
```
Comment (tk0miya) [inline sphinx/environment/__init__.py:352]: why is `filename` optional?

Comment (Zac-HD) [reply sphinx/environment/__init__.py:352]: That's the return value, not the `filename` argument - and `Optional[unicode]` means that it may return a unicode string *or* `None` (if no suffix matches).

Comment (tk0miya) [reply sphinx/environment/__init__.py:352]: oh, sorry. Please forget my comment...

**Comment by Zac-HD** [comment]
Alright, I'm fairly happy with this now.  There are four commits:

- Add tests that doctests report their locations correctly, with real filenames and line numbers
- Tangential refactoring (I was in the area...)
- A nice, principled fix - our reported filename is now based on the block source, rather than the output document
- Paper over the problems with line numbers, by setting to `None` for doctests or subtracting to cancel an off-by-one-error, and making the test weaker.

I'm not delighted by the last bit, but it's still a strict improvement on the status quo and I don't have time to investigate deeply again before the deadline for 1.7.  Assuming that this pull passes tests and review (cc @tk0miya), I'd therefore suggest merging it and leaving a "proper fix" for follow-up work.

**Comment by Zac-HD** [comment]
Ping @tk0miya; I think this is ready to merge now 😄

File: sphinx/ext/doctest.py [hunk 3]
```diff
@@ -366,6 +367,39 @@ def write(self, build_docnames, updated_docnames, method='update'):
...
             self.test_doc(docname, doctree)
 
+    def get_filename_for_node(self, node, docname):
+        # type: (nodes.Node, unicode) -> str
+        """Try to get the file which actually contains the doctest, not the
+        filename of the document it's included in."""
+        try:
+            filename = (
+                path.relpath(node.source, self.env.srcdir)
+                .replace(path.sep, '/')
```
Comment (tk0miya) [inline sphinx/ext/doctest.py:377]: Why do you replace separators by `/`?

File: sphinx/ext/doctest.py [hunk 4]
```diff
@@ -366,6 +367,39 @@ def write(self, build_docnames, updated_docnames, method='update'):
...
+                .rsplit(':docstring of ', maxsplit=1)[0]
+            )
+        except Exception:
+            filename = self.env.doc2path(docname, base=None)
+        if PY2:
+            return filename.encode(fs_encoding)
```
Comment (tk0miya) [inline sphinx/ext/doctest.py:383]: Do you know why is this needed? I think this is not needed, and this should return unicode path string.

File: CHANGES [hunk 5]
```diff
@@ -243,6 +243,7 @@ Bugs fixed
 * #4198: autosummary emits multiple 'autodoc-process-docstring' event. Thanks
   to Joel Nothman.
 * #4081: Warnings and errors colored the same when building
+* #4223: doctest: failing tests reported in wrong file, at wrong line
```
Comment (tk0miya) [inline CHANGES:246]: Please move this to 1.7.0b3 section.

File: tests/test_ext_doctest.py [hunk 6]
```diff
@@ -55,3 +56,27 @@ def test_is_allowed_version():
...
+        'File "index.rst", line 4, in default',
+    ]
+    for location in expected:
+        assert location in failures
+
+
+if PY2:
+    test_reporting_with_autodoc = pytest.mark.xfail(
+        reason='node.source points to document (not filename) under Python 2'
+    )(test_reporting_with_autodoc)
```
Comment (tk0miya) [inline tests/test_ext_doctest.py:82]: Please use decorator style instead.

You can give a condition to `xfail` as an argument.
https://docs.pytest.org/en/2.7.3/skipping.html#mark-a-test-function-as-expected-to-fail

**Review by tk0miya — CHANGES_REQUESTED** [review]
LGTM with nits.

And could you rebase this to 1.7-release branch?
Then I will merge this into 1.7.0 soon.
